### ISSUE TITLE
haskell-interactive-bring process buffer not compatible with org-mode babel ob-haskell session buffer

### ISSUE THREAD

**Issue by stardiviner** [haskell/haskell-mode#1429]
When I execute command `haskell-interactive-bring` will start a process buffer
named `*haskell*` from haskell source code buffer.

When I execute Org-mode Babel haskell src block like this:

```org
#+BEGIN_SRC haskell
1 + 3
#+END_SRC
```
It will start a session buffer named `*haskell*` too.

But the problem is they are not compatible with each other.

If I start haskell process buffer from haskell source code buffer with command
`haskell-interactive-bring` at first, then run the Org-mode haskell babel src
block, then I will get error:

```
org-babel-execute:haskell: Buffer *haskell* does not exist or has no process
```

And in reverse, execute org-mode babel at first to start the session buffer,
then I run command `haskell-interactive-bring` from haskell source code buffer,
I will get another error:

```
Debugger entered--Lisp error: (error "Format specifier doesn’t match argument type")
  format("*%s <%d>*" "haskell" "nil")
  haskell-session-interactive-buffer(((name . "haskell")))
  haskell-interactive-bring()
  funcall-interactively(haskell-interactive-bring)
  call-interactively(haskell-interactive-bring nil nil)
  command-execute(haskell-interactive-bring)
```

**Comment by gracjan** [comment]
Can you provide a PR? As I understand the point is to rename the buffer and make it unique, right?

**Comment by stardiviner** [comment]
No, `ob-haskell` and `haskell-interactive-mode` should both use same `*haskell*` inferior process buffer.
The problem is that one of them started `*haskell*`, but later the other can't recognize it. Report error:
```
Buffer *haskell* does not exist or has no process
```

**Comment by stardiviner** [comment]
Any updated on this?

**Comment by codygman** [comment]
@stardiviner It looks like [ob-haskell](http://orgmode.org/cgit.cgi/org-mode.git/tree/lisp/ob-haskell.el?id=790f5b4d3a898f00bda38a7ea09b8cb51e801a45) will need to be updated to use haskell-interactive-mode instead of inf-haskell.

**Comment by codygman** [comment]
So in ob-haskell I think run-haskell is failing. run-haskell is created by declare-function which appears to start the entire mode or something, I'm not really sure to be honest. Maybe tomorrow I can step through it in a debugger.

**Comment by codygman** [comment]
I also just ran into this issue I believe:

[org-babel--get-vars not defined in orgmode<=8.3.4](https://github.com/pope/ob-go/pull/11)

**Comment by codygman** [comment]
Just for more information, here is my info:

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.1.50.2
- Spacemacs: 0.200.5
- Spacemacs branch: master (rev. 664ba6a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(rust csv yaml auto-completion emacs-lisp git org
      (shell :variables shell-default-height 30 shell-default-position 'bottom)
      syntax-checking markdown haskell purescript javascript clojure scala go)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GCONF GSETTINGS NOTIFY LIBSELINUX GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11

I'm feeling uneasy about getting the issue above since I have the latest org-mode. What I'll do is delete all of my packages and reinstall, then see if I can replicate that error.

**Comment by codygman** [comment]
Neither org-20161224 nor org-plus-contrib-20161224 seem to contain org-babel-get-header or org-babel--get-vars, so we can't use the same fix they did. Actually no, org-babel-get-header seems to be there but I can't find it with grep so I guess it is generated somehow.

The fix they did in the ob-go should work.

**Comment by gracjan** [comment]
Default Haskell session name comes from `haskell-session-default-name` which is `haskell` by default. If you change that there no longer will be collision with `ob-haskell`.

Not that it solves the problem for everybody, but at least it removes the collision for now.

**Comment by stardiviner** [comment]
I tried change the function `haskell-session-default-name` definition:
```elisp
(defun haskell-session-default-name ()
    "Generate a default project name for the new project prompt."
    (let ((file (haskell-cabal-find-file)))
      (or (when file
            (downcase (file-name-sans-extension
                       (file-name-nondirectory file))))
          "*haskell*")))
```
But it does not work.

And in new version Org-mode, new error occurs:

Emacs version:

GNU Emacs 25.1.1 (x86_64-unknown-linux-gnu, GTK+ Version 3.22.5)
 of 2016-12-23
 
Org-mode version: 9.0.3


```
main :: IO ()
main = do
  putStrLn "Hello, World!"
"org-babel-haskell-eoe"
Loaded GHCi configuration from /home/stardiviner/.ghci
λ> 
<interactive>:1:1: error:
    • Variable not in scope: main :: IO ()
    • Perhaps you meant ‘min’ (imported from Prelude)
λ> 
<interactive>:2:8: error: Empty 'do' block
λ> Hello, World!
λ> "org-babel-haskell-eoe"
λ> 
```

**Comment by gracjan** [comment]
Hmm, it looks like it tries to load the haskell source code verbatim to GHCi in single line mode. This will not work this way, it has to save the source code to a file and load that.

I'm not sure how ob-haskell works, but it should have a switch to make it write to file.

**Comment by yiufung** [comment]
Is there any update on this?

**Comment by creichert** [comment]
It's been working well for me with Emacs 25 / haskell-mode-20180917.923

**Comment by dustinlacewell** [comment]
Just ran into this.
